tamr_client/dataset/_dataset.py

```diff
@@ -157,3 +157,100 @@ def materialize(session: Session, dataset: Dataset) -> Operation:
 def _materialize_async(session: Session, dataset: Dataset) -> Operation:
     r = session.post(str(dataset.url) + ":refresh",)
     return operation._from_response(dataset.url.instance, r)
+
+
+def delete(session: Session, dataset: Dataset, *, cascade: bool = False):
+    """Deletes an existing dataset
+
+    Sends a deletion request to the Tamr server
+
+    Args:
+        dataset: Existing dataset to delete
+        cascade: Whether to delete all derived datasets as well
+
+    Raises:
+        dataset.NotFound: If no dataset could be found at the specified URL.
+            Corresponds to a 404 HTTP error.
+        requests.HTTPError: If any other HTTP error is encountered.
+    """
+    r = session.delete(str(dataset.url), params={"cascade": cascade})
+    if r.status_code == 404:
+        raise NotFound(str(dataset.url))
+    response.successful(r)
+
+
+def get_all(
+    session: Session,
+    instance: Instance,
+    *,
+    filter: Optional[Union[str, List[str]]] = None,
+) -> Tuple[Dataset, ...]:
+    """Get all datasets from an instance
+
+    Args:
+        instance: Tamr instance from which to get datasets
+        filter: Filter expression, e.g. "externalId==wobbly"
+            Multiple expressions can be passed as a list
+
+    Returns:
+        The datasets retrieved from the instance
+
+    Raises:
+        requests.HTTPError: If an HTTP error is encountered.
+    """
+    url = URL(instance=instance, path="datasets")
+
+    if filter is not None:
+        r = session.get(str(url), params={"filter": filter})
+    else:
+        r = session.get(str(url))
+
+    datasets_json = response.successful(r).json()
+
+    datasets = []
+    for dataset_json in datasets_json:
+        dataset_url = URL(instance=instance, path=dataset_json["relativeId"])
+        dataset = _from_json(dataset_url, dataset_json)
+        datasets.append(dataset)
+    return tuple(datasets)
+
+
+def create(
+    session: Session,
+    instance: Instance,
+    *,
+    name: str,
+    key_attribute_names: Tuple[str, ...],
+    description: Optional[str] = None,
+    external_id: Optional[str] = None,
+) -> Dataset:
+    """Create a dataset in Tamr.
+
+    Args:
+        instance: Tamr instance
+        name: Dataset name
+        key_attribute_names: Dataset primary key attribute names
+        description: Dataset description
+        external_id: External ID of the dataset
+
+    Returns:
+        Dataset created in Tamr
+
+    Raises:
+        requests.HTTPError: If any other HTTP error is encountered.
+    """
+    data = {
+        "name": name,
+        "keyAttributeNames": key_attribute_names,
+        "description": description,
+        "externalId": external_id,
+    }
+
+    dataset_url = URL(instance=instance, path="datasets")
+    r = session.post(url=str(dataset_url), json=data)
+
```

Comment:
For handling "already exists" errors, we can try to match on the response text to distinguish that error from other 400 errors:

```suggestion
if r.status_code == 400 and f'Dataset "{name}" already exists' in r.text:
    raise AlreadyExists(name)
```

Even if the server changes its error codes for "already exists" upstream, we will still at least catch any errors via `response.successful` on the next line.